PythonVisualizations/AdvancedSorting.py

```diff
@@ -0,0 +1,570 @@
+        y0 = self.ARRAY_Y0 - 15
+        y1 = self.ARRAY_Y0 - 40
+
+        b, l, r = self.drawPartition(left, right)
+
+        self.medianOfThree(left, right)
+        a = self.list
+        pivot = a[right]
+        done = left
+        
+        doneArrow = self.createIndex(done, name="done", level=2)
+        pivotArrow = self.createIndex(right, name="pivot", level=3)
+
+        pivotCellObjects = []
+        pivotCellObjects.append(pivotArrow)
+
+        doneCellObjects = []
+        doneCellObjects.append(doneArrow)
+        
+        curArrow = self.createIndex(left, name="cur")
+        
+        callEnviron |= set(doneArrow)
+        callEnviron |= set(pivotArrow)
+        callEnviron |= set(curArrow)
+        callEnviron |= set((b, l, r))
+        
+        # for each position except for the pivot position
+        for cur in range(left, right):
+            # if the value at that position is smaller than the pivot
+            # swap it so it becomes the next value of the done part
+            if a[cur].val <= pivot.val:
+                #a[done], a[cur] = a[cur], a[done]
+                self.swap(done, cur)
+                done += 1                
+                self.moveItemsBy(doneArrow, (self.CELL_WIDTH, 0), sleepTime=0.02)
+
+            self.wait(0.2)
+
+            self.moveItemsBy(curArrow, (self.CELL_WIDTH, 0), sleepTime=0.02)
+
+        # Move the pivot into the correct place
+        #a[done], a[right] = a[right], a[done]
+        self.swap(done, right, aCellObjects=doneCellObjects, bCellObjects=pivotCellObjects)
+        
+        self.cleanUp(callEnviron)
+
+        # At this point, done is the location where the pivot value got placed
+        self.wait(0.1)
+        return done
+    
+    # Wrapper method 
+    def quickSort(self):
+        # Start animation
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+        
+        self.__quickSort(callEnviron)
+        
+        # Finish animation
+        self.cleanUp()
+    
+    def __quickSort(self, callEnviron, left=-1, right=-1):
+        # initialize things if method was called without args
+        if left == -1:
+            left = 0
+            right = len(self.list) - 1
+
+        # there has to be at least two elements
+        if left < right:
+            partition = self.partitionIt(left, right)
+            self.__quickSort(callEnviron, left, partition - 1)
+            self.__quickSort(callEnviron, partition + 1, right)    
+   
+    def makeButtons(self):
+        # get the common buttons from makeButtons() in SortingBase
+        buttons, vcmd = super().makeButtons()
+        quickSortButton = self.addOperation(
+            "Quick Sort", lambda: self.quickSort())
+        randomFillButton = self.addOperation(
+            "Random Fill", lambda: self.randomFill())
+        newButton = self.addOperation(
+            "New", lambda: self.clickNew(), numArguments=1,
+            validationCmd=vcmd)

```

Comment:
Let's shift the random fill and new operations into the `SortingBase` class since they are common among many visualizations.